ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,500 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
-    ASSERT(Process);
-    ASSERT(Process != PsInitialSystemProcess);
+    SIZE_T ReturnedQuotas;
+    SIZE_T UpdatedLimit, FinalLimit;
+
+    /* Sanity checks */
     ASSERT(QuotaType < PsQuotaTypes);
-    ASSERT(Process->QuotaBlock);
+    ASSERT((SSIZE_T)Amount >= 0);
 
-    /* Guard our quota in a spin lock */
+    /* Guard ourselves in a spin lock */
     KeAcquireSpinLock(&PspQuotaLock, &OldIrql);
 
-    if (Process->QuotaUsage[QuotaType] + Amount >
-        Process->QuotaBlock->QuotaEntry[QuotaType].Limit)
+    /* Are we within the bounds of quota limit? */
+    if (QuotaBlock->QuotaEntry[QuotaType].Usage + Amount >
+        QuotaBlock->QuotaEntry[QuotaType].Limit &&
+        QuotaBlock != &PspDefaultQuotaBlock)
     {
-        DPRINT1("Quota exceeded, but ROS will let it slide...\n");
-        KeReleaseSpinLock(&PspQuotaLock, OldIrql);
-        return STATUS_SUCCESS;
-        //return STATUS_QUOTA_EXCEEDED; /* caller raises the exception */
+        /* We aren't... Is this a page file quota charging? */
+        if (QuotaType == PsPageFile)
+        {
+            /* It is, return the appropriate status code */
+            DPRINT1("PspChargeProcessQuotaSpecifiedPool(): Quota amount exceeds the limit on page file quota (limit -- %lu || amount -- %lu)\n",
+                    QuotaBlock->QuotaEntry[QuotaType].Limit, Amount);
+            return STATUS_PAGEFILE_QUOTA_EXCEEDED;
+        }
+
+        /*
+         * This is not a page file charge. What we can do at best
+         * in this scenario is to attempt to raise (expand) the
+         * quota limit charges of the block.
+         */
+        if (!MmRaisePoolQuota(QuotaType,
+                              QuotaBlock->QuotaEntry[QuotaType].Limit,
+                              &UpdatedLimit))
+        {
+            /*
+             * We can't? It could be that we must free
+             * up some resources in order to raise the
+             * limit, which in that case we must return
+             * the excess of quota that hasn't been
+             * returned. If we haven't returned anything
+             * then what we're doing here is futile.
+             * Bail out...
+             */
+            PspReturnExcessQuotas(QuotaType, &ReturnedQuotas);
+            if (ReturnedQuotas == 0)
+            {
+                DPRINT1("PspChargeProcessQuotaSpecifiedPool(): Failed to free some resources in order to raise quota limits...\n");
+                KeReleaseSpinLock(&PspQuotaLock, OldIrql);
+                return STATUS_QUOTA_EXCEEDED;
+            }
+
+            /* Try to raise the quota limits again */
+            MmRaisePoolQuota(QuotaType,
+                             QuotaBlock->QuotaEntry[QuotaType].Limit,
+                             &UpdatedLimit);
+        }
+
+        /* Enforce a new raised limit */
+        FinalLimit = QuotaBlock->QuotaEntry[QuotaType].Limit - UpdatedLimit;
+        InterlockedExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Limit, FinalLimit);
```

Comment:
In non-interlocked terms, this reads as
```
Limit = Limit - UpdatedLimit;
```
I would expect UpdatedLimit to be larger than Limit, so this seems problematic. Shouldn't we in effect be doing something like
```
Limit = UpdatedLimit;
```
?